tests/firedrake/adjoint/test_assemble.py

```diff
@@ -99,7 +95,7 @@ def test_assemble_1_forms_tlm():
 
     Jhat = ReducedFunctional(J, Control(f))
     h = Function(V)
-    h.vector()[:] = rand(h.dof_dset.size)
+    h.dat.data_wo[...] = rand(h.dof_dset.size)
```

Comment:
We should be consistent with how we are querying the array size for `dat.data`.

1. `dat.data.shape`?
2. `h.dof_dset.size`?
3. `V.dim()`?

Reply:
Thanks. The last one is used in quite a few places so even more reason to make it consistent. I think it's only used in serial tests but we don't want people copying unsafe code.

Reply:
done